### PR TITLE
Bats automate test 40 - verify the default credhelper is set correctly

### PR DESCRIPTION
**Pull request by ericpromislow**
Replaces test plan test # 40

I removed the `assert_success` because it seemed redundant -- all we care about is seeing a `rancher-desktop` context

Also the test is failing for me on linux. I'm expecting to see `pass`, but getting `none`. For some reason, the system isn't setting the `credsStore` field to `pass`. Need to look into this.

### CODE REVIEW

File: bats/tests/helpers/vm.bash [hunk 1]
```diff
@@ -74,7 +74,6 @@ container_engine_info() {
 
 docker_context_exists() {
     run docker_exe context ls -q
-    assert_success
```
Comment (jandubois) [inline bats/tests/helpers/vm.bash:77]: Why did you delete this line? The command needs to succeed before we check that it produces the correct output.

Under which conditions can the command exit with a non-0 return code, but still show the right context?

File: bats/tests/registry/creds.bats [hunk 2]
```diff
@@ -1,7 +1,12 @@
+setup_file() {
+    rm -f "$HOME/.docker/config.json"
+}
+
 setup() {
     load '../helpers/load'
     REGISTRY_IMAGE="registry:2.8.1"
     REGISTRY_PORT="5050"
+    CONFIG_FILE="$HOME/.docker/config.json"
```
Comment (jandubois) [inline bats/tests/registry/creds.bats:9]: Let's call it `DOCKER_CONFIG_FILE`

```suggestion
    DOCKER_CONFIG_FILE="$HOME/.docker/config.json"
```

File: bats/tests/registry/creds.bats [hunk 3]
```diff
@@ -103,6 +108,22 @@ EOF
     wait_for_container_engine
 }
 
+@test 'verify credential is set correctly' {
+    verify_default_credStore
+}
+
+verify_default_credStore() {
+    local CREDHELPER_NAME="$(basename "$CRED_HELPER" .ext | sed s/docker-credential-//)"
```
Comment (jandubois) [inline bats/tests/registry/creds.bats:116]: I assume this should strip the Windows `.exe` extension:

```suggestion
    local CREDHELPER_NAME="$(basename "$CRED_HELPER" .exe | sed s/^docker-credential-//)"
```

It doesn't really matter, but maybe also anchor the regex at the beginning of the name.

File: bats/tests/registry/creds.bats [hunk 4]
```diff
@@ -103,6 +108,22 @@ EOF
...
+
+    assert_file_exist "$CONFIG_FILE"
+    if command -v jq; then
```
Comment (jandubois) [inline bats/tests/registry/creds.bats:119]: `jq` is a prerequisite for our BATS tests; we don't need to provide an alternate implementation.

Having a test perform different operations based on the presence/absence of other tools on the system is a potential source of error: maybe the 2 branches are not covering the same functionality?

File: bats/tests/registry/creds.bats [hunk 5]
```diff
@@ -103,6 +108,22 @@ EOF
...
+        assert_equal "$CREDHELPER_NAME" "$(jq -r .credsStore "$CONFIG_FILE")"
```
Comment (jandubois) [inline bats/tests/registry/creds.bats:120]: Write tests in the simplest form possible. Use `run` and `assert_success`, and then the `assert_output` functions to test the actual output. This makes it easier to see exactly what failed, if the test doesn't pass.

```suggestion
        run jq -r .credsStore "$DOCKER_CONFIG_FILE"
        assert_success
        assert_output "$CREDHELPER_NAME"
```

File: bats/tests/registry/creds.bats [hunk 6]
```diff
@@ -219,3 +240,11 @@ EOF
     run bash -c "echo \"$REGISTRY\" | \"$CRED_HELPER\" get"
     refute_output --partial '"Secret":"password"'
 }
+
+@test 'verify the docker desktop setting is replaced' {
```
Comment (jandubois) [inline bats/tests/registry/creds.bats:244]: The title wasn't immediately clear to me, maybe change to something like:

```suggestion
@test 'verify docker-credential-desktop is replaced with default setting' {
```

File: bats/tests/registry/creds.bats [hunk 7]
```diff
@@ -1,7 +1,12 @@
+setup_file() {
+    rm -f "$HOME/.docker/config.json"
```
Comment (jandubois) [inline bats/tests/registry/creds.bats:2]: I wish there was a way to use the `CONFIG_FILE` setting globally, but it is probably not possible for the setup/teardown functions to share settings.

**Review by jandubois — CHANGES_REQUESTED** [review]
Mostly stylist changes, but the `.ext` extension looks like a bug.

File: bats/tests/registry/creds.bats (hunk 3 above)
Comment (ericpromislow) [reply bats/tests/registry/creds.bats:116]: Yes on the `ext/exe` mixup -- I was thinking "strip the extension", and mis-abbreviated it.

File: bats/tests/registry/creds.bats (hunk 4 above)
Comment (ericpromislow) [reply bats/tests/registry/creds.bats:119]: Good to know `jq` is a pre-rec

File: bats/tests/registry/creds.bats (hunk 5 above)
Comment (ericpromislow) [reply bats/tests/registry/creds.bats:120]: Makes sense

File: bats/tests/registry/creds.bats (hunk 6 above)
Comment (ericpromislow) [reply bats/tests/registry/creds.bats:244]: sure

**Review by jandubois — CHANGES_REQUESTED** [review]
I just looked at the testplan, and it seems this is case 41, not 40.

Test 41 asks to `rm -f "$DOCKER_CONFIG_FILE"` before starting Rancher Desktop because the factory reset doesn't remove it.

**Comment by ericpromislow** [comment]
Regarding test plan numbering, I think the tests got renumbered after I created these tests -- doesn't make sense for me to be consistently off by 1 for every single test I automated.  There used to be two tests numbered 20. There are now two *entries* numbered 22, but one of them is empty.

File: bats/tests/registry/creds.bats [hunk 8]
```diff
@@ -1,7 +1,19 @@
...
```
Comment (jandubois) [inline bats/tests/registry/creds.bats:2]: Please move this to the factory-reset test. Logically it belongs together, it is just that factory-reset doesn't delete this file.

**Review by jandubois — CHANGES_REQUESTED** [review]

File: bats/tests/registry/creds.bats (hunk 8 above)
Comment (ericpromislow) [reply bats/tests/registry/creds.bats:2]: Done.

**Review by jandubois — APPROVED** [review]
Thanks, LGTM
